### PR TITLE
APC Power Delivery Overhaul

### PR DESCRIPTION
**Pull request by Kubius**
[STATION SYSTEMS]
## About the PR 

[Now with a design doc!](https://hackmd.io/@Kubius/HkxHidA13)

Overhaul to APC power delivery. The essentials:

- APC power delivery is now handled by the powernet reset. For load reimbursement, total power demand of APCs is compared to available power to derive a percentage, and each APC will have that percentage of its load reimbursed (up to 100%). As a fallback, APCs that somehow lack a powernet will process on their own.
- The add_load proc (used primarily by APCs but also by SMES and SWORD) now checks against the powernet's newload value, and now has a failure state if that load exceeds the actual available power on-grid. It has also been standardized to use the version defined on power_parent in most cases (APCs have a layer to check for a disabled circuit).
- Netexcess power on a powernet is now offered proportionally to all APCs using the new accept_excess proc; if any APCs are fully charged, they won't be counted against that proportion, allowing highly-loaded cells to recharge smoothly without vast amounts of excess power being required. Any power the APCs do not opt to consume is reabsorbed into SMES, as before.
- Viewload (the displayed load shown in power computers) will now show you the full power demand of APCs (even if the grid was not able to fully satisfy that demand) in addition to direct loads, giving a clearer impression of how well the station's power needs are being met.

I've also done an audit of alterations to newload, particularly outside of an add_load proc, to try and ensure the new charging doesn't fluke out; capacitor pathogen was skipped as it's unlikely to return.

- Capacitor pathogen (unused): Unchanged.
- Carousel power: Now uses additive load.
- Power sink: Now verifies its additive load.
- Portable shield generator: Line power is now an additive load. Shield generator will fall back to cell power if line power isn't adequate.
- Manta junction box (unused): Now verifies its additive load.
- Nadir transception array: Internal capacitor charging is now an additive load.
- PTL: Standardized to terminal-based add_load, since it uses a terminal.

Details of each changed file. Bolded entries are the largest change sets.

- code/WorkInProgress/mantaObjects.dm: Adjusted junction box malfunction to use additive newload.
- code/WorkInProgress/nadir_antenna.dm: Adjusted transception array's internal capacitor charging to use additive newload.
- code/WorkInProgress/recycling/conveyor.dm: Adjusted carousel power unit to use additive newload.
- code/datums/powernet_data.dm: Adjusted perapc's description to be more descriptive, and added an APC charge share variable for the post-load recharge distribution phase.
- **code/modules/power/apc.dm:** Reworks power delivery procedure, splitting what was formerly the APC's process() into a two-stage cycle for the powernet to tap into, and adding a new accept_excess proc to recharge from any spare power.
- **code/modules/power/power_parent.dm:** Powernet side of the reworked power delivery procedure. Adjusted the add_load proc to check for the presence of a powernet.
- code/modules/power/pt_laser.dm, code/modules/power/smes.dm, code/modules/power/sword_engine.dm: Adjusted add_load to use capacity-verified terminal-side charging.
- code/obj/machinery.dm: Adjust the construction mode wire_powered system to fall back to conventional power if wire power can't be supplied with additive newload.
- code/obj/machinery/shield_generators/portable_shield_generator.dm: Adjust line-based power consumption to use additive newload, and smoothly fall back to cell-based operation if line power isn't adequate.

## Why's this needed? 

Allows APCs to charge dynamically based on available power, improves the grid's ability to load-balance, fixes some longstanding flaky behavior and makes the overall power ecosystem more responsive.

Some notable fixes, aside from consumption not exceeding available power and  power being able to route where it's needed:

- APCs no longer ignore loads above the amount of charge left in the cell (effectively yielding free power under certain circumstances).
- SMES will now report their power output even if there isn't excess power at the end of the tick (this rarely cropped up previously because of perapc allocation).

Adequate SMES/solar setups will now be able to supply large loads such as cell chargers smoothly, prioritizing it over trickle charging other cells if needed; if it's desired, later calibration of SMES roundstart power on the map side will allow cell and borg chargers to function smoothly for a brief period before Engineers have an opportunity to set up and supply the SMES.

## Changelog 

```changelog
(u)Kubius
(*)APC power delivery has received a substantial overhaul; the power grid will now prioritize supporting heavy loads, and recharging will scale itself to the amount of spare power available.
```

### CODE REVIEW

**Comment by TheTyrant25** [comment]
How does this interact with hotwired devices? Since this has seemingly scope crept into a power overhaul.

**Comment by Kubius** [comment]
To mention that here for more public clarity, anything that directly uses grid power off the net hasn't (as of time of writing) been touched by this; I'll be doing a pass over direct writes to newload to ensure they're correctly set up to use the now accumulative newload.

**Comment by Kubius** [comment]
Temporarily drafting since Leah found a charge bug (a longstanding one, I think?) that needs to be fixed.

File: code/modules/power/power_parent.dm [hunk 1]
```diff
@@ -377,13 +375,25 @@ var/makingpowernetssince = 0
...
+				var/obj/machinery/power/apc/netapc = term.master
+				var/expended = netapc.accept_excess()				// and give them first share of excess power,
+				if(expended) recharge_sum += expended				// letting the power computer know it's appreciated
+
+		for(var/obj/machinery/power/smes/S in nodes)			// find the SMESes in the network
 			S.restore()				// and restore some of the power that was used
 		for(var/obj/machinery/power/sword_engine/SW in nodes)	//Finds the SWORD Engines in the network.
 			SW.restore()				//Restore some of the power that was used.
+
+	viewload = 0.8*viewload + (0.2*load + 0.2*recharge_sum)
```
[the remark this answers is not in the document]
Comment (Kubius) [reply code/modules/power/power_parent.dm:412]: At a glance it looks like that, but those two numbers are different subsections of the overall load (the former is regular node, the latter is the part of netexcess that was distributed to APCs and not reabsorbed into SMES). I've improved comments throughout that proc in cbd0ff4 for some clarity on this and other parts of the functionality.

**Comment by Kubius** [comment]
[be32d12](https://github.com/goonstation/goonstation/pull/13348/commits/be32d1286bc55d695c9aa8a6602de4b9df86ed38) fixes a longstanding bug not introduced in this PR, mentioned in [this comment](https://github.com/goonstation/goonstation/pull/13348#issuecomment-1465439813).

The amount of power that could be consumed from a cell was (and still is) limited to the amount of capacity that cell has left; however, in the old configuration, that limited consumption also applies to the amount the APC attempts to recoup from the grid, capping the area's real power consumption to the amount of power left in the cell. Cell usage and power consumption are no longer intertwined in this manner.

File: _std/machinery.dm [hunk 2]
```diff
@@ -31,8 +31,8 @@
 //recharger stuff
 /// multiplier for watts per tick != cell storage (eg: .002 means if there is a load of 1000 watts, 20 units will be taken from a cell per second)
 #define CELLRATE 0.002
-/// Cap for how fast cells charge, as a percentage-per-tick (.001 means cellcharge is capped to 1% per second)
-#define CHARGELEVEL 0.001
+/// Cap for how fast cells charge, as a percentage-per-tick (.001 means cellcharge is capped to 1% per second, before interval adjustments)
```
[the remark this answers is not in the document]
Comment (Kubius) [reply _std/machinery.dm:34]: Did this in [2989e28](https://github.com/goonstation/goonstation/pull/13348/commits/2989e283645fbc0ea7f30493d46aca58b7247a38).

**Comment by Kubius** [comment]
If there are concerns that the new delivery has sorta "smoothed the life out of power", there are some ways to reintroduce "clunk" in the overhauled system if desired:

- Alter or remove load deference. There are two defines in place, LOAD_DEFERENCE_THRESHOLD and LOAD_DEFERENCE_PORTION, which respectively dictate the charge level required before load deference kicks in and the amount of load that's deferred. Setting the former to 1 will turn off load deference entirely, preventing the grid from temporarily ignoring that deferred load when demand spikes higher.
- Change recharge behavior to always distribute a fixed share of netexcess to each APC, instead of dynamically adjusting it based on charged APCs; this will cause highly-loaded APCs to not charge above their deference threshold if overall power output is at a moderate level.
- Cap the supply each APC can receive in the main load phase to a fixed percentage of total available power; this is a mixed bag in terms of outcomes, as it can end up reintroducing outages in heavily-loaded areas under circumstances of moderate supply but will keep one room from sucking down the entire available supply.

File: code/modules/power/power_parent.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Azrun) [reply code/modules/power/power_parent.dm:412]: That line is providing a weighted average. I think the way it is portrayed now makes it a little harder to see that.
I think just having in the form `A = X*A + Y(b+c)` where you factor out the 20% of the new value might be better.

[the remark this answers is not in the document]
Comment (Kubius) [reply code/modules/power/power_parent.dm]: Ah! Yep, gotcha, makes sense. Swapped to that.

**Comment by Kubius** [comment]
Drafting again to look into a better power distribution procedure.

**Comment by ZeWaka** [comment]
swordfish/leah has indicated they will sponsor this PR for the freeze

**Comment by TobleroneSwordfish** [comment]
Okay after talking with other devs, we've agreed that this is going to take a lot of figuring out and is best left until after the freeze, sorry for the confusion.

**Comment by TobleroneSwordfish** [comment]
Okay, this was testmerged for a few rounds, the main thing I noticed was that it seems to completely break power sinks, to the point where there were four power sinks being just about sustained by the cog1 SMES + solar output.

**Comment by Kubius** [comment]
Power sinks are pretty wack tuning-wise in general, because they charge grid power units against both grid power and cell power. I was thinking that they needed to be changed either alongside or after this, but not sure whether that should be within this PR or atomized into a discrete one.

**Comment by Kubius** [comment]
Opted to do a basic patch in here because I realized not verifying additive load is 100% breaking things; a4baf97 should restore power sinks to their (kinda borked, but working) live state. Rebalance of the drain units I think I'm more inclined to atomize, but up to you.

**Comment by TheTyrant25** [comment]
certified organic or nah?

**Comment by ZeWaka** [comment]
> certified organic or nah?

nah

**Review by ZeWaka — APPROVED** [review]

**Comment by Kubius** [comment]
Making note of a user-mentioned happening during testmerge: The removal of the bug where APCs ignore loads above the amount of charge left in the cell has caused telescience portals to become difficult to sustain, as a portal consumes the entire capacity of a default power cell (this previously made the sustaining cost much lower, but now instead leaves no buffer at all).

This could be addressed by lowering its activation energy a bit (300-400 kW of load during activation instead of 500kW), but I'm leaning towards that being atomized (by me or someone else) unless it's preferred that I bring that change in here.

**Pull request merged**